Launch dev mode of JupyterLab as a subprocess

yuvipanda opened the pull request:
- Launches JupyterLab in dev mode
- Redirects logs to jupyterlab-dev.log

yuvipanda committed:
Set base_url for dev version of lab

yuvipanda committed:
Remove start

Was also missing a shebang

yuvipanda committed:
Copy config file to appropriate path

yuvipanda committed:
Rename config.py to jupyter_notebook_config.py

ian-r-rose commented on binder/jupyter_notebook_config.py:
```diff
@@ -7,6 +7,7 @@
             '--dev-mode',
             '--port={port}',
             '--NotebookApp.token=""',
```
This makes me nervous. Are there any existing examples of managed servers that require some authentication?

yuvipanda replied:
There are two attack vectors here:

1. People accessing this from the browser without auth. This is prevented by jupyter-server-proxy (hopefully, haha!)
2. People accessing this by hitting 'localhost:<blah>' from within the container using any process. There isn't any protection against this, but it maybe ok?

ian-r-rose replied:
Fair enough, I just wanted to flag it as a possible problem. If I try to sniff out the port of this managed server without going through the proxy, could I get unauthenticated access?

yuvipanda replied:
@ian-r-rose not from your browser, no - since there's no direct port access. You can from inside the container (from a kernel or terminal, for example)

ian-r-rose replied:
What about from curl or python requests (from some random computer across the world)?

yuvipanda replied:
@ian-r-rose yeah, those will be treated same as from your browser, and will be stopped by jupyter-server-proxy

yuvipanda committed:
Use correct path for jupyter_notebook_config.py

yuvipanda committed:
Add debug logs

yuvipanda committed:
Bump jupyter-server-proxy version

Includes https://github.com/jupyterhub/jupyter-server-proxy/pull/91

yuvipanda committed:
Try without the extra /

yuvipanda committed:
Turn off dev-mode

See if the problem is dev-mode

yuvipanda committed:
Use absolute URLs for jupyterlab's dev mode

Thanks to @manics' PR: https://github.com/jupyterhub/jupyter-server-proxy/pull/85

ian-r-rose commented:
@yuvipanda, happy to merge this whenever you feel it is ready, or if you want to take it over, that's cool too.

yuvipanda committed:
Disable dns rebinding protection for dev mode notebook

yuvipanda commented:
@ian-r-rose yeah, not quite there yet! It works on my machine locally, but not on mybinder.org yet

yuvipanda committed:
Don't set global jupyter_notebook_config.py file

This is picked up by the dev instance of JupyterLab as well,
possibly leading to infinite recursion

yuvipanda committed:
Explicitly enable serverextension

ian-r-rose commented on binder/postBuild:
```diff
@@ -1,2 +1,5 @@
+jlpm build
+
+# This seems to be explicitly needed with `pip install -e .`
+jupyter serverextension enable jupyterlab --sys-prefix
```
I've run into this with serverextensions before as well. It seems the `conf.d` stuff doesn't get linked right with an editable install.

ian-r-rose replied:
I think a non `-e` install should work fine, however.

yuvipanda committed:
Don't install jupyterlab with -e

Doesn't seem to put templates in the right place?

yuvipanda committed:
Run jlpm before running build?

Maybe this will help?

yuvipanda committed:
Don't build wheels when installing lab

yuvipanda committed:
Use released version of jupyter-server-proxy

yuvipanda committed:
Redirect logs to a file

gnestor commented:
Ok, I see this was merged into ian-r-rose/jupyterlab, not jupyterlab/jupyterlab...

So let's pretend that this was merged upstream, I have an open PR (https://github.com/jupyterlab/jupyterlab/pull/5670) that's based off of this, and I want to try it out in binder. Would I just go to https://mybinder.org/v2/gh/gnestor/jupyterlab/vdom-events and it would launch in dev-mode? That's what the code implies to me...

One thing to figure out: The binder badge on the jupyterlab README links to https://mybinder.org/v2/gh/jupyterlab/jupyterlab-demo/master?urlpath=lab/tree/demo (notice jupyterlab/jupyterlab-demo repo) where all the demo notebooks are. I guess that we could add a step in the postBuild script that clones the jupyterlab-demo repo and a `?urlpath=lab/tree/demo` to the binder URL. 

Are we waiting to submit a PR for this upstream?

ian-r-rose commented:
I was hoping to make a GitHub bot to post links to the right binder. Taking a look at this now...

gnestor commented:
It looks like we can either create a bot that will comment on a PR with a URL to the binder or we can create a _check_ that will look like this (at https://github.com/jupyter/jupyter.github.io/pull/321):

![image](https://user-images.githubusercontent.com/512354/52376562-46af0380-2a17-11e9-9f29-29a0c1b6fa64.png)

ian-r-rose commented:
Ooh, I had not considered a check. How would that work? We really only need a pretty simple URL templating.

ian-r-rose commented:
She works!

![image](https://user-images.githubusercontent.com/5728311/52382793-8849a980-2a2b-11e9-99d1-3667478236b4.png)

https://github.com/ian-r-rose/jupyterlab-dev-mode-bot

gnestor commented:
Wawaweewa!

![](https://media0.giphy.com/media/a0h7sAqON67nO/giphy.gif?cid=3640f6095c5b7bf7536431566ffd7c11)

So I guess that we just need to submit a PR to jupyterlab and then install the bot once it's merged?

ian-r-rose commented:
Yep, that's it! I just submitted the PR.
